test/lint/lint-filenames-permissions.py

```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+This checks that all files in the repository have correct filenames and permissions
+"""
+
+import os
+import re
+import sys
+from subprocess import check_output
+from typing import Optional, NoReturn
+
+CMD_ALL_FILES = "git ls-files --full-name"
+CMD_SOURCE_FILES = 'git ls-files --full-name -- "*.[cC][pP][pP]" "*.[hH]" "*.[pP][yY]" "*.[sS][hH]"'
+CMD_SHEBANG_FILES = "git grep --full-name --line-number -I '^#!'"
+ALLOWED_FILENAME_REGEXP = "^[a-zA-Z0-9/_.@][a-zA-Z0-9/_.@-]*$"
+ALLOWED_SOURCE_FILENAME_REGEXP = "^[a-z0-9_./-]+$"
+ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP = (
+    "^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)"
+)
+ALLOWED_PERMISSION_NON_EXECUTABLES = 644
+ALLOWED_PERMISSION_EXECUTABLES = 755
+ALLOWED_EXECUTABLE_SHEBANG = {
+    "py": [b"#!/usr/bin/env python3"],
+    "sh": [b"#!/usr/bin/env bash", b"#!/bin/sh"],
+}
+
+
+def get_filename_extension(filename: str) -> Optional[str]:
+    """
+    Returns the full file extension for a given filename string.
+    eg:
+    'ci/lint_run_all.sh' -> 'sh'
+    'ci/retry/retry' -> None
+    'contrib/devtools/split-debug.sh.in' -> 'sh.in'
+    """
+    filename_parts = filename.split(os.extsep, 1)
```

Comment:
I think people mostly do something like this https://source.dot.net/#System.Private.CoreLib/Path.cs,467ebc0e33e0820c to find a file extension. That is to look for the first `os.extsep` but not from the beginning of the string (filename) but from the end of the filename.

A counter-example can be `'contrib/devtools/split-debug.my.nice.helper.sh.in'` where you would return `my.nice.helper.sh.in`.

Your code may be perfectly fine for what you want to achieve in this PR, yet I would be hesitant to use `get_filename_extension` name. 

I have no clear suggestion here, letting it up to you to decide. :)

Reply:
Ah yes I originally was using `os.path.splitext(self.file_path)[1].strip(".")` which works similar to the linked implementation (only return the final extension, not everything from the first `.`)

I had changed this as I needed a way to detect `.sh.in` files to exclude them from one of the tests, but you bring up a very good point - with the change I had made it's possible certain files with weird extensions like `split-debug.my.nice.helper.sh` would not actually be detected as a `sh` file, and would not have certain tests enforced.

Ive updated the code to add a new `FileMeta` class, and have changed the logic to use the appropriate version of the extension in each test. Let me know if this makes sense or not :)

Reply:
It looks fine to me. Thanks for updating it. :)